specs/sharding/beacon-chain.md

```diff
@@ -103,786 +87,558 @@ The following values are (non-configurable) constants used throughout the specif
+    zero_poly = G2_SETUP[1].add(G2_SETUP[0].mult(x).neg())
 
-### Misc
+    assert (
+        bls.Pairing(proof, zero_poly)
+        == bls.Pairing(commitment.add(G1_SETUP[0].mult(y).neg), G2_SETUP[0])
+    )
+```
 
-#### `next_power_of_two`
+#### `interpolate_poly`
 
 ```python
-def next_power_of_two(x: int) -> int:
-    return 2 ** ((x - 1).bit_length())
+def interpolate_poly(xs: List[BLSFieldElement], ys: List[BLSFieldElement]):
+    """
+    Lagrange interpolation
+    """
+    # TODO!
 ```
 
-#### `compute_previous_slot`
+#### `verify_kzg_multiproof`
 
 ```python
-def compute_previous_slot(slot: Slot) -> Slot:
-    if slot > 0:
-        return Slot(slot - 1)
-    else:
-        return Slot(0)
+def verify_kzg_multiproof(commitment: KZGCommitment, xs: List[BLSFieldElement], ys: List[BLSFieldElement], proof: KZGCommitment) -> None:
+    zero_poly = elliptic_curve_lincomb(G2_SETUP[:len(xs)], interpolate_poly(xs, [0] * len(ys)))
+    interpolated_poly = elliptic_curve_lincomb(G2_SETUP[:len(xs)], interpolate_poly(xs, ys))
+
+    assert (
+        bls.Pairing(proof, zero_poly)
+        == bls.Pairing(commitment.add(interpolated_poly.neg()), G2_SETUP[0])
+    )
 ```
 
-#### `compute_updated_sample_price`
+#### `verify_degree_proof`
 
 ```python
-def compute_updated_sample_price(prev_price: Gwei, samples_length: uint64, active_shards: uint64) -> Gwei:
-    adjustment_quotient = active_shards * SLOTS_PER_EPOCH * SAMPLE_PRICE_ADJUSTMENT_COEFFICIENT
-    if samples_length > TARGET_SAMPLES_PER_BLOB:
-        delta = max(1, prev_price * (samples_length - TARGET_SAMPLES_PER_BLOB) // TARGET_SAMPLES_PER_BLOB // adjustment_quotient)
-        return min(prev_price + delta, MAX_SAMPLE_PRICE)
-    else:
-        delta = max(1, prev_price * (TARGET_SAMPLES_PER_BLOB - samples_length) // TARGET_SAMPLES_PER_BLOB // adjustment_quotient)
-        return max(prev_price, MIN_SAMPLE_PRICE + delta) - delta
+def verify_degree_proof(commitment: KZGCommitment, degree: uint64, proof: KZGCommitment):
+    """
+    Verifies that the commitment is of polynomial degree <= degree. 
+    """
+
+    assert (
+        bls.Pairing(proof, G2_SETUP[0])
+        == bls.Pairing(commitment, G2_SETUP[-degree - 1])
+    )
 ```
 
-#### `compute_committee_source_epoch`
+#### `block_to_field_elements`
 
 ```python
-def compute_committee_source_epoch(epoch: Epoch, period: uint64) -> Epoch:
+def block_to_field_elements(block: bytes) -> List[BLSFieldElement]:
     """
-    Return the source epoch for computing the committee.
+    Slices a block into 31 byte chunks that can fit into field elements
     """
-    source_epoch = Epoch(epoch - epoch % period)
-    if source_epoch >= period:
-        source_epoch -= period  # `period` epochs lookahead
-    return source_epoch
+    sliced_block = [block[i:i + 31] for i in range(0, len(bytes), 31)]
+    return [BLSFieldElement(int.from_bytes(x, "little")) for x in sliced_block]
```

Comment:
Is 31 to guarantee being in the scalar field of BLS12-381?
Is there a size bound on the length of the list?

Reply:
Yes, we can only store 31 bytes in a BLS field element.